spec/ga_spec.t

```diff
@@ -0,0 +1,298 @@
+      assert.equal(a, GA4.vector(102,112,66,-92) + GA4.trivector(24,25,-6,3))
+    end
+  end)
+
+  it('should multiply bivectors with bivectors', terra()
+    do
+      var a = GA2.bivector(2)*GA2.bivector(3)
+      assert.equal(a, GA2.scalar(-6))
+    end
+
+    do
+      var a = GA3.bivector(3,4,5)*GA3.bivector(6,7,8)
+      assert.equal(a, GA3.scalar(-86) + GA3.bivector(3,-6,3))
+    end
+
+    do
+      var a = GA4.bivector(3,4,5,6,7,8)*GA4.bivector(9,10,11,12,13,14)
+      var c = GA4.scalar(-397) + GA4.bivector(12,0,12,-48,0,24) + GA4.vector4(118)
+      assert.truthy(a == c)
+    end
+  end)
+
+  it('should correctly rotate vectors', terra()
+    var PI = 3.14159f
+
+    do
+      var i = GA3.bivector(1,0,1):normalize()
+      var v = (GA3.exp(-i*(PI/4))*GA3.vector(1,1,1))*GA3.exp(i*(PI/4))
+      assert.truthy((v - GA3.vector(0.29289, 0, 1.70711)):magnitude() <= 0.0001f)
+    end
+
+    do
+      var i = GA3.bivector(1,0,0)
+      var v = (GA3.exp(-i*(PI/4))*GA3.vector(1,0,1))*(Math.cos(PI/4) + i*Math.sin(PI/4))
+      --var v = ((Math.cos(PI/4) + (-i)*Math.sin(PI/4))*GA3.vector(1,0,1))*(Math.cos(PI/4) + i*Math.sin(PI/4))
+      assert.truthy((v - GA3.vector(0, 1, 1)):magnitude() <= 0.00001f)
+    end
+  end)
+
+  it('should calculate the dot product correctly', terra()
+    var f = GA3.bivector(1,2,3)
+    -- This must be EXACTLY equal, because it SHOULD be the exact same sequence of float calculations
+    assert.equal(f:magnitude(), Math.sqrt(Math.fabs_32(f:dot(f))))
+  end)
+
+  it('should calculate the wedge product correctly', terra()
+    var f = GA3.bivector(1,2,3)
+    assert.equal(f^f, [GA3.multivector({3,5,6})]{array(0.f,0.f,0.f)})
+    assert.equal(GA3.vector(1,0,0)^GA3.vector(0,1,0), GA3.bivector(1,0,0))
+  end)
+
+  it('should allow accessing components via x/y/z/w', terra()
+    var v = GA4.scalar(1) + GA4.vector(2,3,4,5) + GA4.bivector(6,7,8,9,10,11) + GA4.trivector(12,13,14,15) + GA4.vector4(16)
+    assert.equal(v.x,2)
+    assert.equal(v.y,3)
+    assert.equal(v.z,4)
+    assert.equal(v.w,5)
+    assert.equal(v.xy,6)
+    assert.equal(v.xyz,12)
+    assert.equal(v.xyzw,16)
+  end)
+
+  it('should do grade projection correctly', terra()
+    var v = GA4.scalar(1) + GA4.vector(2,3,4,5) + GA4.bivector(6,7,8,9,10,11) + GA4.trivector(12,13,14,15) + GA4.vector4(16)
+    assert.equal(v:gradeproj(0), GA4.scalar(1))
+    assert.equal(v:gradeproj(1), GA4.vector(2,3,4,5))
+    assert.equal(v:gradeproj(2), GA4.bivector(6,7,8,9,10,11))
+    assert.equal(v:gradeproj(3), GA4.trivector(12,13,14,15))
+    assert.equal(v:gradeproj(4), GA4.vector4(16))
+  end)
+
+  it('should project and reject correctly', terra()
+    var v = GA3.vector(1,1,1)
+    var B = GA3.bivector(1,0,0) -- project on to xy-plane
+    assert.equal(v:project(B), GA3.vector(1,1,0))
+    assert.equal(v:reject(B), GA3.vector(0,0,1))
+  end)
+
+  it('should calculate volume correctly', terra()
+    var verts : GA3.vector_t[8]
```

Comment:
You can't use array() with the vector construction methods, but if you really want to initialize the list, you could also do this via:

```
var verts = array(GA3.vector_t{1,1,0}, GA3.vector_t{2,1,0}, GA3.vector_t{2,2,0}, GA3.vector_t{2,1,0}, etc...)
```